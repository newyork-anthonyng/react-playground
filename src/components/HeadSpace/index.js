import React, { Component } from "react";
import T from "prop-types";

class HeadSpace extends Component {
  constructor() {
    super();

    // keep track of y-scroll positions
    this.previousScrollPosition = 0;
    this.currentScrollPosition = 0;

    // the height of our header container
    this.startOffset = 0;

    this.isTicking = false;

    this.state = {
      isHidden: false,
      isFixed: false
    };
  }

  componentDidMount() {
    window.addEventListener("scroll", this.handleScroll);

    this.startOffset = this.container.offsetHeight;
  }

  componentWillUnmount() {
    window.removeEventListener("scroll", this.handleScroll);
  }

  handleScroll = () => {
    this.currentScrollPosition = window.pageYOffset;

    this.requestTick();
  };

  requestTick = () => {
    // https://www.html5rocks.com/en/tutorials/speed/animations/
    // Use requestAnimationFrame to throttle scroll events
    if (!this.isTicking) {
      requestAnimationFrame(this.update);
    }
    this.isTicking = true;
  };

  update = () => {
    this.isTicking = false;

    if (this.currentScrollPosition <= 0) {
      this.resetHeader();
    } else if (this.props.showAtBottom && this.isAtEndOfPage()) {
      this.fixHeader();
    } else if (this.isScrolledGreaterThanTolerance()) {
      const isScrollingDown =
        this.currentScrollPosition > this.previousScrollPosition;

      if (isScrollingDown) {
        this.hideHeader();
      } else {
        this.fixHeader();
      }
    }

    this.previousScrollPosition = this.currentScrollPosition;
  };

  isScrolledGreaterThanTolerance = () => {
    return (
      this.currentScrollPosition > this.startOffset &&
      Math.abs(this.currentScrollPosition - this.previousScrollPosition) >=
        this.props.scrollTolerance
    );
  };

  isAtEndOfPage = () => {
    return (
      window.innerHeight + this.currentScrollPosition >=
      document.body.offsetHeight
    );
  };

  resetHeader = () => {
    this.setState({
      isHidden: false,
      isFixed: false
    });
  };

  hideHeader = () => {
    this.setState({ isHidden: true });
  };

  fixHeader = () => {
    this.setState({
      isFixed: true,
      isHidden: false
    });
  };

  render() {
    const containerStyle = {
      position: this.state.isFixed ? "fixed" : "absolute",
      transform: this.state.isHidden
        ? `translate3d(0, -${this.startOffset}px, 0)`
        : "translate3d(0, 0, 0)",
      top: 0,
      left: 0,
      right: 0,
      transition: "transform 0.3s"
    };
    const placeholderStyle = {
      height: `${this.startOffset}px`
    };

    return (
      <div style={placeholderStyle}>
        <div
          ref={container => {
            this.container = container;
          }}
          style={containerStyle}
        >
          {this.props.children}
        </div>
      </div>
    );
  }
}

HeadSpace.defaultProps = {
  scrollTolerance: 5,
  showAtBottom: true
};

HeadSpace.propTypes = {
  children: T.node.isRequired,
  scrollTolerance: T.number.isRequired,
  showAtBottom: T.bool.isRequired
};

export default HeadSpace;
