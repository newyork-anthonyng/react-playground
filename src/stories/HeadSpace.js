import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, number, boolean, text } from '@storybook/addon-knobs';
import glamorous from 'glamorous';

import HeadSpace from '../components/HeadSpace';

const Container = glamorous.div({
  height: '3000vh',
});
const MyHeader = glamorous.h1({
  backgroundColor: 'tomato',
  textAlign: 'center',
  margin: '0',
});

storiesOf('HeadSpace', module)
  .addDecorator(withKnobs)
  .add('With knobs', () => {
    return (
      <Container>
        <HeadSpace>
          <MyHeader>Hello World</MyHeader>
        </HeadSpace>
        <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam et fringilla urna, ut scelerisque tellus. Proin aliquet nisi nunc. Fusce ac mauris non ligula posuere molestie id non magna. Cras congue nunc dui. Mauris mattis efficitur leo nec interdum. Phasellus sem leo, egestas sit amet molestie id, scelerisque id mi. Nam nec pellentesque diam. Vestibulum nulla libero, dignissim sed lectus at, pulvinar rutrum tellus. Etiam vehicula suscipit aliquet. Nullam ac porttitor ex. Donec ultricies bibendum nibh a pharetra. Donec eget orci id justo pellentesque viverra nec vitae metus. Curabitur eget viverra felis, vel mollis sem. Praesent venenatis magna vitae elit dignissim scelerisque. Suspendisse ut purus purus. Integer vel orci cursus, pretium elit eget, iaculis quam.</p>
      </Container>
    );
  })
