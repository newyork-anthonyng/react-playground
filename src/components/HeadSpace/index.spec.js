import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import HeadSpace from './';

// timer is needed to mock asynchronous nature of window.requestAnimationFrame
jest.useFakeTimers()

// Mock window events
const map = {};
window.addEventListener = jest.fn((type, cb) => {
  map[type] = cb;
});
window.removeEventListener = jest.fn(type => {
  map[type] = undefined;
});
window.requestAnimationFrame = jest.fn(cb => {
  window.setTimeout(cb, 0);
});
window.innerHeight = 200; // size of viewport

// window.document.body.offsetHeight is read-only
Object.defineProperty(window.document.body, 'offsetHeight', {
  value: 1000,
});

const defaultProps = {
  scrollTolerance: 10,
  showAtBottom: true,
  children: (<h1>Child</h1>),
};

describe('Event listeners', () => {
  it('should add scroll event listener on mount', () => {
    mount(
      <HeadSpace {...defaultProps} />
    );

    expect(window.addEventListener.mock.calls).toMatchSnapshot();
  });

  it('should remove scroll event listener on unmount', () => {
    const wrapper = mount(
      <HeadSpace {...defaultProps} />
    );
    wrapper.unmount();

    expect(window.removeEventListener).toHaveBeenCalledTimes(1);
    expect(window.removeEventListener.mock.calls).toMatchSnapshot();
  });
});

it('should render children correctly', () => {
  const wrapper = shallow(
    <HeadSpace {...defaultProps}>
      <h1>I am the header!</h1>
    </HeadSpace>
  );

  expect(toJSON(wrapper)).toMatchSnapshot();
});

describe('When scrolling', () => {
  it('should reset header when scrolling to the top of the page', () => {
    const wrapper = mount(
      <HeadSpace {...defaultProps} />
    );
    window.pageYOffset = 0;
    map.scroll();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  describe('When at end of page', () => {
    it('should fix header if showing at bottom', () => {
      const wrapper = mount(
        <HeadSpace
          {...defaultProps}
          showAtBottom
        />
      );
      /*
       * 800 (scroll position) + 200 (size of viewport)
       * >== 1000 (size of HTML content)
       */
      window.pageYOffset = 800;
      map.scroll();
      jest.runOnlyPendingTimers();

      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('should not fix header if not showing at bottom', () => {
      const wrapper = mount(
        <HeadSpace
          {...defaultProps}
          showAtBottom={false}
        />
      );
      /*
       * 800 (scroll position) + 200 (size of viewport)
       * >== 1000 (size of HTML content)
       */
      window.pageYOffset = 800;
      map.scroll();
      jest.runOnlyPendingTimers();

      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  it('should not do anything if scrolling less than tolerance', () => {
    const scrollTolerance = 42;
    const wrapper = mount(
      <HeadSpace
        {...defaultProps}
        scrollTolerance={scrollTolerance}
      />
    );

    window.pageYOffset = scrollTolerance - 2;
    map.scroll();

    window.pageYOffset = window.pageYOffset + (scrollTolerance - 2);
    map.scroll();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  describe('When scrolling greater than tolerance level', () => {
    it('should hide header if scrolling down', () => {
      const scrollTolerance = 42;
      const wrapper = mount(
        <HeadSpace
          {...defaultProps}
          scrollTolerance={scrollTolerance}
        />
      );

      window.pageYOffset = scrollTolerance + 2;
      map.scroll();
      jest.runOnlyPendingTimers();

      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('should fix header if scrolling up', () => {
      const scrollTolerance = 42;
      const wrapper = mount(
        <HeadSpace
          {...defaultProps}
          scrollTolerance={scrollTolerance}
        />
      );
      window.pageYOffset = scrollTolerance + 100;
      map.scroll();
      jest.runOnlyPendingTimers();

      // scroll back up
      window.pageYOffset = window.pageYOffset - (scrollTolerance + 2);
      map.scroll();
      jest.runOnlyPendingTimers();

      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });
});
