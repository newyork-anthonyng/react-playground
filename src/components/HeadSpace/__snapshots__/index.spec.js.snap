// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event listeners should add scroll event listener on mount 1`] = `
Array [
  Array [
    "scroll",
    [Function],
    false,
  ],
  Array [
    "scroll",
    [Function],
  ],
]
`;

exports[`Event listeners should remove scroll event listener on unmount 1`] = `
Array [
  Array [
    "scroll",
    [Function],
  ],
]
`;

exports[`When scrolling When at end of page should fix header if showing at bottom 1`] = `
<HeadSpace
  scrollTolerance={10}
  showAtBottom={true}
>
  <div
    style={
      Object {
        "height": "0px",
      }
    }
  >
    <div
      style={
        Object {
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
          "transform": "translate3d(0, 0, 0)",
          "transition": "transform 0.3s",
        }
      }
    >
      <h1>
        Child
      </h1>
    </div>
  </div>
</HeadSpace>
`;

exports[`When scrolling When at end of page should not fix header if not showing at bottom 1`] = `
<HeadSpace
  scrollTolerance={10}
  showAtBottom={false}
>
  <div
    style={
      Object {
        "height": "0px",
      }
    }
  >
    <div
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": "translate3d(0, -0px, 0)",
          "transition": "transform 0.3s",
        }
      }
    >
      <h1>
        Child
      </h1>
    </div>
  </div>
</HeadSpace>
`;

exports[`When scrolling When scrolling greater than tolerance level should fix header if scrolling up 1`] = `
<HeadSpace
  scrollTolerance={42}
  showAtBottom={true}
>
  <div
    style={
      Object {
        "height": "0px",
      }
    }
  >
    <div
      style={
        Object {
          "left": 0,
          "position": "fixed",
          "right": 0,
          "top": 0,
          "transform": "translate3d(0, 0, 0)",
          "transition": "transform 0.3s",
        }
      }
    >
      <h1>
        Child
      </h1>
    </div>
  </div>
</HeadSpace>
`;

exports[`When scrolling When scrolling greater than tolerance level should hide header if scrolling down 1`] = `
<HeadSpace
  scrollTolerance={42}
  showAtBottom={true}
>
  <div
    style={
      Object {
        "height": "0px",
      }
    }
  >
    <div
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": "translate3d(0, -0px, 0)",
          "transition": "transform 0.3s",
        }
      }
    >
      <h1>
        Child
      </h1>
    </div>
  </div>
</HeadSpace>
`;

exports[`When scrolling should not do anything if scrolling less than tolerance 1`] = `
<HeadSpace
  scrollTolerance={42}
  showAtBottom={true}
>
  <div
    style={
      Object {
        "height": "0px",
      }
    }
  >
    <div
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": "translate3d(0, 0, 0)",
          "transition": "transform 0.3s",
        }
      }
    >
      <h1>
        Child
      </h1>
    </div>
  </div>
</HeadSpace>
`;

exports[`When scrolling should reset header when scrolling to the top of the page 1`] = `
<HeadSpace
  scrollTolerance={10}
  showAtBottom={true}
>
  <div
    style={
      Object {
        "height": "0px",
      }
    }
  >
    <div
      style={
        Object {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": "translate3d(0, 0, 0)",
          "transition": "transform 0.3s",
        }
      }
    >
      <h1>
        Child
      </h1>
    </div>
  </div>
</HeadSpace>
`;

exports[`should render children correctly 1`] = `
<div
  style={
    Object {
      "height": "0px",
    }
  }
>
  <div
    style={
      Object {
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": "translate3d(0, 0, 0)",
        "transition": "transform 0.3s",
      }
    }
  >
    <h1>
      I am the header!
    </h1>
  </div>
</div>
`;
